resources/js/register-components.tsx

```diff
@@ -0,0 +1,153 @@
+import { LandingNews } from 'components/landing-news';
+import MainNotificationIcon from 'components/main-notification-icon';
+import QuickSearchButton from 'components/quick-search-button';
+import RankingFilter from 'components/ranking-filter';
+import { RankingType } from 'components/ranking-filter';
+import RankingSelectOptions from 'components/ranking-select-options';
+import SpotlightSelectOptions from 'components/spotlight-select-options';
+import { UserCard } from 'components/user-card';
+import { UserCardStore } from 'components/user-card-store';
+import { startListening, UserCardTooltip } from 'components/user-card-tooltip';
+import { UserCards } from 'components/user-cards';
+import { WikiSearch } from 'components/wiki-search';
+import GameMode from 'interfaces/game-mode';
+import { keyBy } from 'lodash';
+import { observable } from 'mobx';
+import { deletedUser } from 'models/user';
+import NotificationWidget from 'notification-widget/main';
+import core from 'osu-core-singleton';
+import QuickSearch from 'quick-search/main';
+import QuickSearchWorker from 'quick-search/worker';
+import * as React from 'react';
+import { parseJson, parseJsonNullable } from 'utils/json';
+
+function reqJson<T>(input: string|undefined): T {
+  // This will throw when input is missing and thus parsing empty string.
+  return JSON.parse(input ?? '') as T;
+}
+
+function reqStr(input: string|undefined) {
+  if (input == null) {
+    throw new Error('unexpected undefined value');
+  }
+
+  return input;
+}
+
+core.reactTurbolinks.register('countdownTimer', (container) => (
+  <CountdownTimer deadline={reqStr(container.dataset.deadline)} />
+));
+
+core.reactTurbolinks.register('blockButton', (container) => (
+  <BlockButton userId={parseInt(reqStr(container.dataset.target), 10)} />
+));
+
+core.reactTurbolinks.register('beatmap-discussion-events', () => {
+  const props: BeatmapsetEventsProps = {
+    events: parseJson('json-events'),
+    mode: 'list',
+    users: keyBy(parseJson('json-users'), 'id'),
+  };
+
+  // TODO: move to store?
+  // eslint-disable-next-line id-blacklist
+  props.users.null = props.users.undefined = deletedUser.toJson();
+
+  return <BeatmapsetEvents {...props} />;
+});
+
+core.reactTurbolinks.register('beatmapset-panel', (container) => {
+  const props: BeatmapsetPanelProps = reqJson(container.dataset.beatmapsetPanel);
+
+  return <BeatmapsetPanel {...observable(props)} />;
+});
+
+core.reactTurbolinks.register('ranking-select-options', () => (
+  <RankingSelectOptions {...parseJson('json-ranking-select-options')} />
+));
+
+core.reactTurbolinks.register('spotlight-select-options', () => (
+  <SpotlightSelectOptions {...parseJson('json-spotlight-select-options')} />
+));
+
+core.reactTurbolinks.register('comments', (container) => {
+  const props: CommentsManagerProps = {
+    component: Comments,
+    ...reqJson<Omit<CommentsManagerProps, 'component'>>(container.dataset.props),
```

Comment:
properties being assigned here should come after the spread, otherwise it's misleading as `reqJson<Omit..` doesn't actually exclude the omitted property from the parsing.

Reply:
but it's more of indication the json doesn't contain component 🤔

Reply:
that's why it should be the other way around because it'll still parse `component` if it's in the json overriding the already set `component` when it's expected `component` to be always `Comment`